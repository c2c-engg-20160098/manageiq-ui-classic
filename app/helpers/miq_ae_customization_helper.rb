module MiqAeCustomizationHelper
  def dialog_id_action
    url = request.parameters
    if url[:id].present?
      {:id => @record.id.to_s, :action => 'edit'}
    elsif url[:copy].present?
      {:id => url[:copy], :action => 'copy'}
    else
      {:id => '', :action => 'new'}
    end
  end
end
