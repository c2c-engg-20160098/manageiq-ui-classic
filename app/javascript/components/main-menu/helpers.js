export const getHrefByType = (type, href, id) => {
  switch (type) {
    case 'big_iframe':
      return `/dashboard/iframe?id=${id}`;
    case 'modal':
      return undefined;
    default:
      return href;
  }
};

export const getTargetByType = type => (type === 'new_window' ? '_blank' : '_self');

export const getItemId = id => `menu_item_${id}`;

export const getSectionId = id => `menu_section_${id}`;

export const getIdByCategory = (isSection, id) => (isSection ? getSectionId(id) : getItemId(id));

export const handleUnsavedChanges = (type) => {
  if (type === 'modal') {
    return sendDataWithRx({ type: 'showAboutModal' });
  }
  return window.miqCheckForChanges();
};
