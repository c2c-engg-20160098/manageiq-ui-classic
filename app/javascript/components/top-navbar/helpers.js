export const showAboutModal = (e) => {
  e.preventDefault();
  return sendDataWithRx({ type: 'showAboutModal' });
};
