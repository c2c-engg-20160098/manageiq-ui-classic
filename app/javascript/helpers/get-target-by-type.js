const getTargetByType = type => (type === 'new_window' ? '_blank' : '_self');

export default getTargetByType;
