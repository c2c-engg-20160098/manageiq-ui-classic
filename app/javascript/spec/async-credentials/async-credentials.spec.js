import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { FieldProviderComponent as FieldProvider } from '../helpers/fieldProvider';
import AsyncCredentials from '../../components/async-credentials/async-credentials';

describe('Async credentials component', () => {
  let initialProps;
  const DummyComponent = ({ isDisabled, ...props }) => <input {...props} />;
  beforeEach(() => {
    initialProps = {
      name: 'async-wrapper',
      fields: [{
        name: 'foo',
      }, {
        name: 'bar',
      }],
      FieldProvider,
      asyncValidate: jest.fn(),
      formOptions: {
        renderForm: fields => fields.map(field => <DummyComponent key={field.name} {...field} />),
        getState: () => ({
          values: {
            foo: 'value-foo',
            bar: 'value-bar',
            nonAsync: 'non-async',
          },
        }),
        change: jest.fn(),
      },
    };
  });

  it('should render correctly', () => {
    const wrapper = shallow(<AsyncCredentials {...initialProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call async validation function on button click and set valid state to true', async(done) => {
    const asyncValidate = jest.fn().mockReturnValue(new Promise(resolve => resolve('Ok')));
    const change = jest.fn();
    const wrapper = mount(<AsyncCredentials {...initialProps} formOptions={{ ...initialProps.formOptions, change }} asyncValidate={asyncValidate} />);

    await act(async() => {
      wrapper.find('button').simulate('click');
    });
    expect(asyncValidate).toHaveBeenCalledWith({
      foo: 'value-foo',
      bar: 'value-bar',
      nonAsync: 'non-async',
    }, ['foo', 'bar']);
    expect(change).toHaveBeenCalledWith('async-wrapper', true);
    done();
  });

  it('should call async validation function on button click and set valid state to false', async(done) => {
    const asyncValidate = jest.fn().mockReturnValue(new Promise((_resolve, reject) => reject('Validation failed'))); // eslint-disable-line prefer-promise-reject-errors
    const change = jest.fn();
    const wrapper = mount(<AsyncCredentials {...initialProps} formOptions={{ ...initialProps.formOptions, change }} asyncValidate={asyncValidate} />);

    await act(async() => {
      wrapper.find('button').simulate('click');
    });
    expect(asyncValidate).toHaveBeenCalledWith({
      foo: 'value-foo',
      bar: 'value-bar',
      nonAsync: 'non-async',
    }, ['foo', 'bar']);
    expect(change).toHaveBeenCalledWith('async-wrapper', false);
    done();
  });

  it('should correctly set invalid state after input change', () => {
    const change = jest.fn();
    const getStateMock = jest.fn()
      .mockReturnValueOnce({
        values: {
          foo: 'value-foo',
          bar: 'value-bar',
          nonAsync: 'non-async',
        },
      })
      .mockReturnValue({
        values: {
          foo: 'changed-value',
          bar: 'value-bar',
          nonAsync: 'non-async',
        },
      });

    const expectedChangeCalls = [
      ['foo', 'changed-value'],
      ['async-wrapper', false],
    ];
    const wrapper = mount(<AsyncCredentials
      {...initialProps}
      formOptions={{
        ...initialProps.formOptions,
        change,
        getState: getStateMock,
      }}
    />);
    wrapper.find('input').first().simulate('change', { target: { value: 'changed-value' } });

    expect(change).toHaveBeenCalledTimes(2);
    expect(change.mock.calls).toEqual(expectedChangeCalls);
  });

  it('should correctly set valid state after input change if passed initial values', () => {
    const change = jest.fn();
    const getStateMock = jest.fn()
      .mockReturnValue({
        values: {
          foo: 'changed-value',
          bar: 'value-bar',
          nonAsync: 'non-async',
        },
      });

    const expectedChangeCalls = [
      ['foo', 'changed-value'],
      ['async-wrapper', undefined],
    ];
    const wrapper = mount(<AsyncCredentials
      {...initialProps}
      formOptions={{
        ...initialProps.formOptions,
        change,
        getState: getStateMock,
      }}
    />);
    wrapper.find('input').first().simulate('change', { target: { value: 'changed-value' } });

    expect(change).toHaveBeenCalledTimes(2);
    expect(change.mock.calls).toEqual(expectedChangeCalls);
  });

  it('should correctly set valid state after input change to initial empty values', () => {
    const change = jest.fn();
    const getStateMock = jest.fn()
      .mockReturnValueOnce({
        values: {},
      })
      .mockReturnValueOnce({
        values: {
          foo: 'changed-value',
        },
      })
      .mockReturnValueOnce({
        values: {
          foo: 'changed-value',
        },
      })
      .mockReturnValue({
        values: {},
      });

    const expectedChangeCalls = [
      ['foo', 'changed-value'],
      ['async-wrapper', false],
      ['foo', ''],
      ['async-wrapper', undefined],
    ];
    const wrapper = mount(<AsyncCredentials
      {...initialProps}
      formOptions={{
        ...initialProps.formOptions,
        change,
        getState: getStateMock,
      }}
    />);
    wrapper.find('input').first().simulate('change', { target: { value: 'changed-value' } });
    wrapper.find('input').first().simulate('change', { target: { value: '' } });
    expect(change).toHaveBeenCalledTimes(4);
    expect(change.mock.calls).toEqual(expectedChangeCalls);
  });
});
