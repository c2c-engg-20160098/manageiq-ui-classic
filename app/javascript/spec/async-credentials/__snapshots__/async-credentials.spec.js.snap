// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async credentials component should render correctly 1`] = `
<ContextProvider
  value="async-wrapper"
>
  <DummyComponent
    isDisabled={false}
    key="foo"
    name="foo"
    onChange={[Function]}
  />
  <DummyComponent
    isDisabled={false}
    key="bar"
    name="bar"
    onChange={[Function]}
  />
  <FieldProviderComponent
    initialValue={false}
    name="async-wrapper"
    validate={[Function]}
  >
    <Component />
  </FieldProviderComponent>
</ContextProvider>
`;
